compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -633,6 +638,29 @@ impl<'a> Parser<'a> {
         self.parse_expr_unary(lo, UnOp::Not)
     }
 
+    /// Parse `box expr` - this syntax has been removed, but we still parse this
+    /// for now to provide an automated way to fix usages of it
+    fn parse_expr_box(&mut self, lo: Span) -> PResult<'a, (Span, ExprKind)> {
+        let (span, expr) = self.parse_expr_prefix_common(lo)?;
+        let code = self.sess.source_map().span_to_snippet(span.with_lo(lo.hi())).unwrap();
+        self.sess.emit_err(errors::BoxSyntaxRemoved { span, code: code.trim() });
+        // So typechecking works, parse `box <expr>` as `::std::boxed::Box::new(expr)`
+        let path = Path {
+            span: DUMMY_SP,
```

Comment:
There's a span right above, you should use it. Also the path thing doesn't really work on `no_std`+`alloc`, but whatever, it's fine, you can't really do it properly here. Might lead to slightly confusing errors for users of that, but it should be clear enough.

Reply:
I used the dummy span here since I thought it didn't really make sense to attach a span to a path which is being produced from out of nowhere, but I can change that
As for the error, maybe a note that "parsing this as `::std::boxed::Box::new(expr)`" would help?

Reply:
Yeah ideally you'd do `::alloc::boxed::Box` as that works both in `no_std` and in default contexts.

Both `alloc` and `std` can be shadowed, so yes this isn't perfect but I don't think it *needs* to be: it just has to work well enough to support rustfix in most cases. Shadowing alloc or std should be quite rare.

Even if the error message if shadowing is present is a bit weird, the "box is removed" thingy is still printed so in the worst case you'd get another weird error. This exists with recovery stuff all the time, for example you can get errors or warnings about stuff being unused because the code that uses it is after some thingy that broke.

Reply:
`alloc` isn't available in default contexts without `extern crate alloc`, is it?

Reply:
@clubby789 oh good point about `alloc`, I wasn't aware of that. But you are right, I've just tried it out.